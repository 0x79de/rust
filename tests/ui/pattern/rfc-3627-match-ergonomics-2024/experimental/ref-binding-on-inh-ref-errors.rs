//@ edition: 2024
//@ revisions: classic structural
//@[classic] run-pass
//! Tests for errors from binding with `ref x` under a by-ref default binding mode. These can't be
//! in the same body as tests for other errors, since they're emitted during THIR construction.
#![allow(incomplete_features)]
#![cfg_attr(classic, feature(ref_pat_eat_one_layer_2024))]
#![cfg_attr(structural, feature(ref_pat_eat_one_layer_2024_structural))]

pub fn main() {
    let [&ref x] = &[&0];
    //[structural]~^ ERROR: this pattern relies on behavior which may change in edition 2024
    //[structural]~| cannot override to bind by-reference when that is the implicit default
    #[cfg(classic)] let _: &&u32 = x;

    let [&ref x] = &[&mut 0];
    //[structural]~^ ERROR: this pattern relies on behavior which may change in edition 2024
    //[structural]~| cannot override to bind by-reference when that is the implicit default
    #[cfg(classic)] let _: &&mut u32 = x;

    let [&ref x] = &mut [&0];
    //[structural]~^ ERROR: this pattern relies on behavior which may change in edition 2024
    //[structural]~| cannot override to bind by-reference when that is the implicit default
    #[cfg(classic)] let _: &&u32 = x;

    let [&ref x] = &mut [&mut 0];
    //[structural]~^ ERROR: this pattern relies on behavior which may change in edition 2024
    //[structural]~| cannot override to bind by-reference when that is the implicit default
    #[cfg(classic)] let _: &&mut u32 = x;

    let [&mut ref x] = &mut [&mut 0];
    //[structural]~^ ERROR: this pattern relies on behavior which may change in edition 2024
    //[structural]~| cannot override to bind by-reference when that is the implicit default
    #[cfg(classic)] let _: &&mut u32 = x;

    let [&mut ref mut x] = &mut [&mut 0];
    //[structural]~^ ERROR: this pattern relies on behavior which may change in edition 2024
    //[structural]~| cannot override to bind by-reference when that is the implicit default
    #[cfg(classic)] let _: &mut &mut u32 = x;
}
