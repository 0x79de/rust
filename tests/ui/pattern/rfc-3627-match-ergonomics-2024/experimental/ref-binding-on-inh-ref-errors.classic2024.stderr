error[E0308]: mismatched types
  --> $DIR/ref-binding-on-inh-ref-errors.rs:54:10
   |
LL |     let [&mut ref x] = &[&mut 0];
   |          ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&ref x] = &[&mut 0];
   |          ~

error: binding modifiers may only be written when the default binding mode is `move`
  --> $DIR/ref-binding-on-inh-ref-errors.rs:67:10
   |
LL |     let [ref mut x] = &[0];
   |          ^^^^^^^ default binding mode is `ref`
   |
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2024/match-ergonomics.html>
help: make the implied reference pattern explicit
   |
LL |     let &[ref mut x] = &[0];
   |         +

error[E0596]: cannot borrow data in a `&` reference as mutable
  --> $DIR/ref-binding-on-inh-ref-errors.rs:67:10
   |
LL |     let [ref mut x] = &[0];
   |          ^^^^^^^^^ cannot borrow as mutable

error: binding modifiers may only be written when the default binding mode is `move`
  --> $DIR/ref-binding-on-inh-ref-errors.rs:75:10
   |
LL |     let [ref x] = &[0];
   |          ^^^ default binding mode is `ref`
   |
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2024/match-ergonomics.html>
help: make the implied reference pattern explicit
   |
LL |     let &[ref x] = &[0];
   |         +

error: binding modifiers may only be written when the default binding mode is `move`
  --> $DIR/ref-binding-on-inh-ref-errors.rs:79:10
   |
LL |     let [ref x] = &mut [0];
   |          ^^^ default binding mode is `ref mut`
   |
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2024/match-ergonomics.html>
help: make the implied reference pattern explicit
   |
LL |     let &mut [ref x] = &mut [0];
   |         ++++

error: binding modifiers may only be written when the default binding mode is `move`
  --> $DIR/ref-binding-on-inh-ref-errors.rs:83:10
   |
LL |     let [ref mut x] = &mut [0];
   |          ^^^^^^^ default binding mode is `ref mut`
   |
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2024/match-ergonomics.html>
help: make the implied reference pattern explicit
   |
LL |     let &mut [ref mut x] = &mut [0];
   |         ++++

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0308, E0596.
For more information about an error, try `rustc --explain E0308`.
